### PR TITLE
Add async ServerOptionsSelectionCallback UseHttps overload

### PR DESCRIPTION
**Pull request by halter73**
The changes to HttpsConnectionMiddleware to support this UseHttps overload were made in #24286. I didn't want to slow down that PR with public API considerations, but now is the time to review the API.

This not only allows async SNI configuration, but allows configuring nearly any part of SslServerAuthenticationOptions asynchronously after getting the ClientHello.

Initially, I was a little concerned that this doesn't respect configuration made via ConfigureHttpsDefaults or EndpointDefaults in config, but then I noticed that this not unique to this UseHttps overload. `UseHttps(HttpsConnectionAdapterOptions)` has this same limitation, because unlike `UseHttps(Action<HttpsConnectionAdapterOptions>)` and like this new overload, there are no incoming options to pre-populate with defaults.

Addresses #20981

@davidfowl @Tratcher

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs [hunk 1]
```diff
@@ -235,7 +234,37 @@ public static ListenOptions UseHttps(this ListenOptions listenOptions, HttpsConn
...
-        /// <param name="state">State for the <see cref="ServerOptionsSelectionCallback" />.</param>
+        /// <param name="state">State for the <paramref name="httpsOptionsCallback"/>.</param>
+        /// <returns>The <see cref="ListenOptions"/>.</returns>
+        public static ListenOptions UseHttps(this ListenOptions listenOptions, ServerOptionsSelectionCallback httpsOptionsCallback, object state)
+        {
+            return listenOptions.UseHttps(httpsOptionsCallback, state, HttpsConnectionAdapterOptions.DefaultHandshakeTimeout);
+        }
+
+        /// <summary>
+        /// Configure Kestrel to use HTTPS.
```
Comment (pranavkm) [inline src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:245]: Add a doc comment that says this no longer uses defaults.

File: src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs [hunk 2]
```diff
@@ -235,7 +234,37 @@ public static ListenOptions UseHttps(this ListenOptions listenOptions, HttpsConn
         /// </summary>
         /// <param name="listenOptions">The <see cref="ListenOptions"/> to configure.</param>
         /// <param name="httpsOptionsCallback">Callback to configure HTTPS options.</param>
-        /// <param name="state">State for the <see cref="ServerOptionsSelectionCallback" />.</param>
+        /// <param name="state">State for the <paramref name="httpsOptionsCallback"/>.</param>
+        /// <returns>The <see cref="ListenOptions"/>.</returns>
+        public static ListenOptions UseHttps(this ListenOptions listenOptions, ServerOptionsSelectionCallback httpsOptionsCallback, object state)
```
Comment (pranavkm) [inline src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:239]:
```suggestion
        public static ListenOptions UseHttps(this ListenOptions listenOptions, ServerOptionsSelectionCallback serverOptionsSelectionCallback, object state)
```

File: src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs [hunk 3]
```diff
@@ -235,7 +234,37 @@ public static ListenOptions UseHttps(this ListenOptions listenOptions, HttpsConn
...
+
+        /// <summary>
+        /// Configure Kestrel to use HTTPS.
+        /// </summary>
+        /// <param name="listenOptions">The <see cref="ListenOptions"/> to configure.</param>
+        /// <param name="httpsOptionsCallback">Callback to configure HTTPS options.</param>
+        /// <param name="state">State for the <paramref name="httpsOptionsCallback"/>.</param>
+        /// <param name="handshakeTimeout">Specifies the maximum amount of time allowed for the TLS/SSL handshake. This must be positive and finite.</param>
+        /// <returns>The <see cref="ListenOptions"/>.</returns>
+        public static ListenOptions UseHttps(this ListenOptions listenOptions, ServerOptionsSelectionCallback httpsOptionsCallback, object state, TimeSpan handshakeTimeout)
```
Comment (pranavkm) [inline src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:252]:
```suggestion
        public static ListenOptions UseHttps(this ListenOptions listenOptions, ServerOptionsSelectionCallback serverOptionsSelectionCallback, object state, TimeSpan handshakeTimeout)
```

**Review by jkotalik — APPROVED** [review]

**Review by Tratcher — APPROVED** [review]
No tests or samples?

**Comment by Pilchie** [comment]
Approved for RC1 pending CI completion if merged before 10am Pacific on 2020-09-01.

**Comment by davidfowl** [comment]
I'm not here, I'm on vacation but @halter73 can you take over my PR https://github.com/dotnet/aspnetcore/pull/24935

**Comment by nitinag** [comment]
Thanks for adding this. We're using `ServerOptionsSelectionCallback` and `SslStreamCertificateContext` to enable hosting thousands of domains and selecting the right ssl certificates with Kestrel. Though X509Certificate2 on Windows does have temporary file creation issues that don't disappear with dispose that we have to deal with.

I had to spend some time to figure this out and to get it working with HTTP/2. Posting some sample code here for the benefit of others trying to use `ServerOptionsSelectionCallback`, for inclusion in documentation, and if anything with the sample code triggers changes/improvements.

We needed `ConfigureAlpn` to enable HTTP/2. Maybe this method should be made public in `HttpsConnectionMiddleware` or called internally in `HttpsConnectionMiddleware` for us?

Also would be great to have access to `connectionContext` so we can abort the connection if we didn't find a matching certificate or for any other reasons. If there is a best practice there on how to drop the connection when a matching certificate is not found, would be glad to hear it. Right now, we just return null as the certificate and let Kestrel drop the connection.

```c#
listenOptions.UseHttps(ServerOptionsSelectionCallback, state: listenOptions.Protocols);

ValueTask<SslServerAuthenticationOptions> ServerOptionsSelectionCallback(
	SslStream stream,
	SslClientHelloInfo clientHelloInfo,
	object state,
	CancellationToken cancellationToken)
{
	var serverOptions = new SslServerAuthenticationOptions
	{
		EnabledSslProtocols = SslProtocols.Tls12 | SslProtocols.Tls13,
		CertificateRevocationCheckMode = X509RevocationMode.NoCheck,
		ServerCertificateContext = SslStreamCertificateContext.Create(certificate, intermediateCertificates, offline: true)
	};

	//ConfigureAlpn / Enable Http2
	var httpProtocols = (HttpProtocols)state;
	ConfigureAlpn(serverOptions, httpProtocols);

	return new ValueTask<SslServerAuthenticationOptions>(serverOptions);
}

//Copy of method from:
//https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs
void ConfigureAlpn(SslServerAuthenticationOptions serverOptions, HttpProtocols httpProtocols)
{
	serverOptions.ApplicationProtocols = new List<SslApplicationProtocol>();

	// This is order sensitive
	if ((httpProtocols & HttpProtocols.Http2) != 0)
	{
		serverOptions.ApplicationProtocols.Add(SslApplicationProtocol.Http2);
		// https://tools.ietf.org/html/rfc7540#section-9.2.1
		serverOptions.AllowRenegotiation = false;
	}

	if ((httpProtocols & HttpProtocols.Http1) != 0)
	{
		serverOptions.ApplicationProtocols.Add(SslApplicationProtocol.Http11);
	}
}
```

**Pull request merged**